File: LCD.py
# 160x144 pixels

class LCD(object):
    def __init__(self, mem):
        self.mem = mem
        self.counter = 0
        mem.put(0xff44, 0) # LCD Y pos

    def __str__(self):
        return "LCD\nLY: %d" % self.mem.get(0xff44)

    def step(self):
        self.counter += 1
        if self.counter >= 100:
            self.counter = 0
            LY = self.mem.get(0xff44)
            LY += 1
            if LY == 144:
                self.mem.put(0xff0f, self.mem.get(0xff0f) | 0x01) # VBLANK
            if LY > 153:
                LY = 0
            self.mem.put(0xff44, LY)

File: src/LCD.cpp
#include "LCD.h"

const unsigned short IF = 0xFF0F;
const unsigned short LCDC = 0xFF40;
const unsigned short LY = 0xFF44;

LCD::LCD(Memory *memory) : mem(memory), cycles(0) {
  ly = mem->getByte(LY);
}

void
LCD::step() {
  cycles += mem->cycles;

  while(cycles > 114) {
    cycles -= 114;
    ly++;
    if(ly == 144) {
      // Vblank Interrupt Request
      mem->putByte(IF, mem->getByte(IF) | 0x01);
    }
    if(ly >= 154) {
      ly = 0;
    }
    if(ly == 0) {
      // Turn off Vblank Interrupt Request
      mem->putByte(IF, mem->getByte(IF) & ~0x01);
    }
    mem->putByte(LY, ly);
  }
}
